File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:


        my_dict = {}

        def robHouse(i):

            #Base Case
            if i > len(nums) - 1:
                return 0

            #Recursive WITH DP
            if i in my_dict:
                return my_dict[i]

            #Rob
            robIt = nums[i] + robHouse(i + 2)

            #Skip
            skipIt = robHouse(i + 1)
            
            my_dict[i] = max(robIt, skipIt)

            return my_dict[i]


        return robHouse(0)
        
        

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        
        int startZero = dfs(0, nums);
        int startOne = dfs(1, nums);

        maxRes = std::max(startZero, startOne);

        return maxRes;
    }
    
    int dfs(int i, vector<int>& nums)
    {
        int curr = 0;
        //Base Case //BOUNDS
        if (i >= nums.size())
        {
            return 0;
        }

        if (dp.count(i))
        {
            return dp[i];
        }

        //Rob the house
        curr += nums[i];
        
        int take = curr + dfs(i + 2, nums);
        int skip = dfs(i + 1, nums);

        dp[i] = std::max(take, skip);

        return dp[i];
    }

private:
    int maxRes = INT_MIN;
    unordered_map<int, int> dp;
};